### DISCUSSION TITLE
ESP32 sml additonal smart meter no readings

### DISCUSSION THREAD

**Discussion by hvullu** [arendst/Tasmota#20634] [Q&A] [answered]
Hi there,
I got a esp32 with sml enabled firmware (dev 13.0.2.0) . Connected two smartmeter sdm720 on a serial port each (r16/t17 and r25/t26). Each has address 1 on individual bus (converter rs485 to ttl). Everything works fine so far for over many month.
Now I need to add a third meter (same make and model) changed its modbus-address to 2 and connected it to the bus #2 (r25/t26).
Strange things happen. I don't get the readings of both devices. If I only use device with address #1 I get the readings,
choosing the settings for device #2 I get the readings, but not both simultaneously.
I'm obviously not able to track down my error.
Below the original sml script (works) and then the extended script (just get the last meter readings 'CEE', the 'WMP' are shown all with zero readings).
Maybe some wise additional pair of eyes can identify my error.
Thanks 4 ur help, have a nice one.

(labels are in german - sorry for that)

The original working for month:

`>D
a="1,010404ffffffff@"
b="2,010404ffffffff@"
>B
->sensor53 r
>M 2
+1,16,mN1,0,9600,Haus,17,15,01040000,01040002,01040004,01040006,01040008,0104000a,0104001e,01040020,01040022,0104000c,0104000e,01040010,01040034,01040046,01040156
+2,25,mN1,0,9600,WMP,26,15,0104000c,0104000e,01040010,01040034,01040156

%a%i0:1,Spannung L1,V,U_l1,2
%a%i1:1,Spannung L2,V,U_l2,2
%a%i2:1,Spannung L3,V,U_l3,2
%a%i3:1,Strom L1,A,I_l1,2
%a%i4:1,Strom L2,A,I_l2,2
%a%i5:1,Strom L3,A,I_l3,2
%a%i6:1,Cos ϕ L1,,λ_l1,2
%a%i7:1,Cos ϕ L2,,λ_l2,2
%a%i8:1,Cos ϕ L3,,λ_l3,2
%a%i9:1,Leistung L1,W,P_l1,2
%a%i10:1,Leistung L2,W,P_l2,2
%a%i11:1,Leistung L3,W,P_l3,2
%a%i12:1,akt. Leistung ∑,W,P_sum,2
%a%i13:1,Systemfrequenz,Hz,f,2
%a%i14:1,Gesamt kum.,kWh,W_cum,2

%b%i0:1,Leistung L1,W,P_l1,2
%b%i1:1,Leistung L2,W,P_l2,2
%b%i2:1,Leistung L3,W,P_l3,2
%b%i3:1,akt. Leistung ∑,W,P_sum,2
%b%i4:1,Gesamt kum.,kWh,W_cum,2
`

The extended not working, just CEE is shown, for WMP are alle values are empty

`>D
; zeitfaktor parameter 8 hw abhaengig - anderer esp andere zeit, testen!
a="1,010404ffffffff@"
b="2,010404ffffffff@"
c="3,020404ffffffff@"
>B
->sensor53 r
>M 3
+1,16,mN1,0,9600,Haus,17,15,01040000,01040002,01040004,01040006,01040008,0104000a,0104001e,01040020,01040022,0104000c,0104000e,01040010,01040034,01040046,01040156
+2,25,mN1,0,9600,WMP,26,30,0104000c,0104000e,01040010,01040034,01040156
+3,25,mN1,0,9600,CEE,26,30,0204000c,0204000e,02040010,02040034,02040156

%a%i0:1,Spannung L1,V,U_l1,2
%a%i1:1,Spannung L2,V,U_l2,2
%a%i2:1,Spannung L3,V,U_l3,2
%a%i3:1,Strom L1,A,I_l1,2
%a%i4:1,Strom L2,A,I_l2,2
%a%i5:1,Strom L3,A,I_l3,2
%a%i6:1,Cos ϕ L1,,λ_l1,2
%a%i7:1,Cos ϕ L2,,λ_l2,2
%a%i8:1,Cos ϕ L3,,λ_l3,2
%a%i9:1,Leistung L1,W,P_l1,2
%a%i10:1,Leistung L2,W,P_l2,2
%a%i11:1,Leistung L3,W,P_l3,2
%a%i12:1,akt. Leistung ∑,W,P_sum,2
%a%i13:1,Systemfrequenz,Hz,f,2
%a%i14:1,Gesamt kum.,kWh,W_cum,2

%b%i0:1,Leistung L1,W,P_l1,2
%b%i1:1,Leistung L2,W,P_l2,2
%b%i2:1,Leistung L3,W,P_l3,2
%b%i3:1,akt. Leistung ∑,W,P_sum,2
%b%i4:1,Gesamt kum.,kWh,W_cum,2

%c%i0:1,Leistung L1,W,P_l1,2
%c%i1:1,Leistung L2,W,P_l2,2
%c%i2:1,Leistung L3,W,P_l3,2
%c%i3:1,akt. Leistung ∑,W,P_sum,2
%c%i4:1,Gesamt kum.,kWh,W_cum,2`

**Comment by gemu2015** [comment]
from core 2.0.12 above ESP32 UART management is broken. (only 2 UARTS of 3 usable) known  bug and will be solved in core 2.0.15
use older Tasmota version or define 2 meters on the same bus. (see example in docs)

**Reply by Noschvie** [reply]
There are 2 meters connected to "r25/t26", this should work, isn't it?

**Reply by gemu2015** [reply]
ok, i see only 2 UARTS used, but then the definition is wrong. a meter is defined per bus. you must define 2 meters only.
see example in docs how to define 2 meters on same bus.

**Reply by hvullu** [reply, marked as answer]
Thanks,
not a straight solution but a hint to it. I was heavily misleaded by the discription of the M parameter. "increase per meter" that is wrong as soon as the meter joins a bus with another meter! So better would be to state "increase M" per bus. To get the readings of additional meters the registers of both must be placed in the M line. Something like:

`...
b="2,010404ffffffff@"
c="2,020404ffffffff@"
...
>M 2
+1,16,mN1,0,9600,B1,17,15,01040000,01040002, ..............
; use registers of BOTH meters !!
+2,25,mN1,0,9600,B2,26,15,0104000c,0104000e,01040010,01040034,01040156,0204000c,0204000e,02040010,02040034,02040156
...
%b%i0:1,WMP L1,W,WMP_P_l1,2
%b%i1:1,WMP L2,W,WMP_P_l2,2
%b%i2:1,WMP L3,W,WMP_P_l3,2
%b%i3:1,WMP L∑,W,WMP_P_sum,2
%b%i4:1,WMP kum.,kWh,WMP_W_cum,2

%c%i5:1,CEE L1,W,CEE_P_l1,2
%c%i6:1,CEE L2,W,CEE_P_l2,2
%c%i7:1,CEE L3,W,CEE_P_l3,2
%c%i8:1,CEE L∑,W,CEE_P_sum,2
%c%i9:1,CEE kum.,kWh,CEE_W_cum,2
`
